[GSoC] Adding Discrete-Time TransferFunction

leomanga opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
This PR lays the groundwork for extending the lti module to discrete time. It implements the `TransferFunctionBase` class, which is inherited by both `TransferFunction` and `DiscreteTransferFunction`. It also extends the ability to create interconnections between discrete-time systems.

Note: This PR is a clone of #28115. See that PR for the complete history.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.control
  * Added a new class for discrete time transfer functions.
  * Added a general method for creating both discrete-time and continuous-time transfer functions
<!-- END RELEASE NOTES -->

leomanga committed:
Squash from disc-time-systems

moorepants commented on sympy/physics/control/lti.py:
```diff
@@ -1230,224 +1258,958 @@ def get_asymptotic_stability_conditions(self, cancel_poles_zeros = False):
 
-    __rtruediv__ = __truediv__
+    @classmethod
+    def forward_diff(cls, cont_tf, sampling_time, var):
```
`from_forward_diff`?

moorepants commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -239,7 +268,7 @@ def test_TransferFunction_functions():
 
     # evaluate the transfer function at particular frequencies.
-    assert tf1.eval_frequency(wn) == wn**2/(wn**2 + 4*wn - 5) + 2*wn/(wn**2 + 4*wn - 5) - 3/(wn**2 + 4*wn - 5)
+    assert tf1.eval_frequency(wn) == wn/(wn + 5) + 3/(wn + 5)
```
Rather than changing tests, it would be better to test symbolically like so:

```python
assert sm.simplify(tf1.eval_frequency(wn) - (wn**2/(wn**2 + 4*wn - 5) + 2*wn/(wn**2 + 4*wn - 5) - 3/(wn**2 + 4*wn - 5)))
```

leomanga committed:
Fix typo

Co-authored-by: Jason K. Moore <moorepants@gmail.com>

moorepants commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -1378,92 +2922,172 @@ def test_TransferFunctionMatrix_functions():
     tf = TransferFunction(1, a*s+b, s)
     numZ, denZ = gbt(tf, T, 0.5)
     # discretized transfer function with coefs from tf.gbt()
-    tf_test_bilinear = TransferFunction(s * numZ[0] + numZ[1], s * denZ[0] + denZ[1], s)
```
Are tests removed here? It isn't so clear to me.

moorepants commented on sympy/printing/latex.py:
```diff
@@ -2692,11 +2698,17 @@ def _print_MIMOFeedback(self, expr):
+        else:
+            print_mat = r"%s_k \text{, sampling time: } {%s}" % (mat,
+                                                           expr.sampling_time)
+        return print_mat
```
I don't see any tests for the new printers. Where are testing the latex and pretty printing?

moorepants commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -485,6 +529,538 @@ def test_TransferFunction_is_biproper():
+    assert G4.expand() == DiscreteTransferFunction(a0*s**s - b0*p**p, g, p, 2.4)
+    assert G4.expand() != DiscreteTransferFunction(a0*s**s - b0*p**p, g, p)
+
+    # purely symbolic polynomials.
```
Change comment to "testing that subs works", I assumed it was testing the construction of a purely symbolic transfer function but maybe you do that somewhere else.

leomanga committed:
physics.control: Add tests in pretty and latex printing.
Change name of classmethods in DiscreteTransferFunction class and improve tests.
Remove DiscreteStateSpace functionalities

leomanga committed:
Merge branch 'master' into disc-time-systems-clear

moorepants commented:
```
 sympy.physics.control.lti.DiscreteTransferFunction._eval_rewrite_as_DiscreteStateSpace 
File "/home/runner/work/sympy/sympy/sympy/physics/control/lti.py", line 2188, in sympy.physics.control.lti.DiscreteTransferFunction._eval_rewrite_as_DiscreteStateSpace
Failed example:
    dtf.rewrite(DiscreteStateSpace)
Exception raised:
    Traceback (most recent call last):
      File "/opt/hostedtoolcache/Python/3.13.6/x64/lib/python3.13/doctest.py", line 1395, in __run
        exec(compile(example.source, filename, "single",
        ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                     compileflags, True), test.globs)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "<doctest sympy.physics.control.lti.DiscreteTransferFunction._eval_rewrite_as_DiscreteStateSpace[3]>", line 1, in <module>
        dtf.rewrite(DiscreteStateSpace)
        ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
      File "/home/runner/work/sympy/sympy/sympy/core/basic.py", line 2087, in rewrite
        return self._rewrite(pattern, rule, method, **hints)
               ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/runner/work/sympy/sympy/sympy/core/basic.py", line 2099, in _rewrite
        rewritten = meth(*args, **hints)
      File "/home/runner/work/sympy/sympy/sympy/physics/control/lti.py", line 2199, in _eval_rewrite_as_DiscreteStateSpace
        return DiscreteStateSpace(A, B, C, D, self.sampling_time)
      File "/home/runner/work/sympy/sympy/sympy/physics/control/lti.py", line 6402, in __new__
        raise NotImplementedError("DiscreteStateSpace is not implemented yet. ")
    NotImplementedError: DiscreteStateSpace is not implemented yet. 

===== tests finished: 4642 passed, 1 failed, 30 skipped, in 218.94 seconds =====
```

Best to just remove that class. We shouldn't add not implemented to the code base.

leomanga committed:
physics.control: Remove method _eval_rewrite_as_DiscreteStateSpace

leomanga committed:
Merge branch 'disc-time-systems-clear' of github.com:leomanga/sympy into disc-time-systems-clear

leomanga committed:
physics.control: Remove DiscreteStateSpace from imports and tests
